File: 02.py
with open("2.txt") as f:
    cmds = [x for x in f.readlines()]

    cur_depth = 0
    cur_pos = 0
    for c in cmds:
        if c.startswith('forward'):
            cur_pos += int(c[len('forward') + 1:])
        elif c.startswith('up'):
            cur_depth -= int(c[len('up') + 1:])
        elif c.startswith('down'):
            cur_depth += int(c[len('down') + 1:])

    print('part1', cur_depth * cur_pos)

    cur_depth = 0
    cur_pos = 0
    cur_aim = 0
    for c in cmds:
        if c.startswith('forward'):
            x = int(c[len('forward') + 1:])
            cur_pos += x
            cur_depth += cur_aim * x
        elif c.startswith('up'):
            cur_aim -= int(c[len('up') + 1:])
        elif c.startswith('down'):
            cur_aim += int(c[len('down') + 1:])

    print('part2', cur_depth * cur_pos)

File: 02.clj
; read file
(def lines
  (map 
    (fn[line] 
      (let [cmd (clojure.string/split line #" ")
            num (Integer/parseInt (cmd 1))]
        [(cmd 0) num]))
    (clojure.string/split-lines (slurp "2.txt"))))

(defn do-part1 
  ([lines] (do-part1 lines 0 0))
  ([lines x d]
    (if (empty? lines)
      (* x d)
      (let [line (first lines)
            num (line 1)]
        (case (line 0)
          "forward" (recur (rest lines) (+ x num) d)
          "down" (recur (rest lines) x (+ d num))
          "up" (recur (rest lines) x (- d num)))))))

(do-part1 lines)
(time (println "part1" (do-part1 lines)))
  
(defn do-part2
  ([lines] (do-part2 lines 0 0 0))
  ([lines x d a]
    (if (empty? lines)
      (* x d)
      (let [line (first lines)
            num (line 1)]
        (case (line 0)
          "forward" (recur (rest lines) (+ x num) (+ d (* a num)) a)
          "down" (recur (rest lines) x d (+ a num))
          "up" (recur (rest lines) x d (- a num)))))))

(do-part2 lines)
(time (println "part2" (do-part2 lines)))